### PR TITLE
feat: Make relationship soft deletable [DHIS2-11320]

### PR DESCRIPTION
**Pull request by enricocolasante**
This PR implements soft deletion for relationship.
Few considerations:

- Relationship are not updatable. If the client tries to update a relationship, so a relationship with a uid that is already present in the DB, that relationship is ignored (validation is skipped) and a warning is returned.
- A new validation is implemented to check that the client is not creating a duplicate relationship. 2 relationships are duplicated if they have same relationship type, from and to. If the relationship type is bidirectional, even relationships with from and to swapped are considered duplicates.
- Some e2e tests were changed to reflect the change in the contract. Before creating a duplicate relationship was just returning a warning, now it is throwing an error.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java [hunk 1]
```diff
@@ -73,21 +74,39 @@ public void validateRelationship( ValidationErrorReporter reporter, Relationship
...
             validateReferences( reporter, relationship, relationship.getFrom() );
             validateReferences( reporter, relationship, relationship.getTo() );
         }
 
     }
 
+    private void validateDuplication( ValidationErrorReporter reporter, Relationship relationship,
+        TrackerBundle bundle )
+    {
+        org.hisp.dhis.relationship.Relationship relationshipUsingKey = bundle.getPreheat()
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java:96]: name it `existingRelationship` like you did in the other hook

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java [hunk 2]
```diff
@@ -123,8 +126,8 @@ public void setUp()
         when( bundle.getProgramInstance( ENROLLMENT_UID ) ).thenReturn( getEnrollment() );
         when( bundle.getProgramStageInstance( SOFT_DELETED_EVENT_UID ) ).thenReturn( getSoftDeletedEvent() );
         when( bundle.getProgramStageInstance( EVENT_UID ) ).thenReturn( getEvent() );
-        when( preheat.getRelationship( getPayloadRelationship() ) )
-            .thenReturn( getRelationship() );
+        when( preheat.getRelationship( RELATIONSHIP_UID ) ).thenReturn( getRelationship() );
+        when( preheat.getRelationship( SOFT_DELETED_RELATIONSHIP_UID ) ).thenReturn( softDeletedRelationship() );
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java:130]: Could you please add the mock to the test needing it.

I try to remove `@MockitoSettings( strictness = Strictness.LENIENT )` when I can in our tests. This helps us keeping them clean (only needed mocks/stubs) in the long run.

If you have the energy and time you could 🧽 this test class and get rid of `@MockitoSettings( strictness = Strictness.LENIENT )` 😋 Or in a later PR, so we get this merged quicker

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java [hunk 3]
```diff
@@ -505,6 +508,22 @@ void verifyRelationshipValidationFailsWhenIsDeleteAndRelationshipIsNotPresent()
         hasTrackerError( reporter, E4016, RELATIONSHIP, rel.getUid() );
     }
 
+    @Test
+    void verifyRelationshipValidationFailsWhenIsSoftDeleted()
+    {
+        // given
+        Relationship rel = Relationship.builder()
+            .relationship( SOFT_DELETED_RELATIONSHIP_UID )
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java:516]: I prefer keeping the test fixtures local if possible. So I would do create a relationship using `softDeletedRelationship(someUid)` here, pass that into the `rel` and the mock.

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java [hunk 4]
```diff
@@ -505,6 +508,22 @@ void verifyRelationshipValidationFailsWhenIsDeleteAndRelationshipIsNotPresent()
         hasTrackerError( reporter, E4016, RELATIONSHIP, rel.getUid() );
     }
 
+    @Test
+    void verifyRelationshipValidationFailsWhenIsSoftDeleted()
+    {
+        // given
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java:514]: I don't think we need these comments. Your test is very readable 😄 with the arrange, act, assert groups.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java [hunk 5]
```diff
@@ -311,6 +311,19 @@ public List<Relationship> getByUids( List<String> uids )
         }
     }
 
+    @Override
+    protected void preProcessPredicates( CriteriaBuilder builder,
+        List<Function<Root<Relationship>, Predicate>> predicates )
+    {
+        predicates.add( root -> builder.equal( root.get( "deleted" ), false ) );
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java:318]: does this mean all methods in here get this predicate added?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java [hunk 6]
```diff
@@ -311,6 +311,19 @@ public List<Relationship> getByUids( List<String> uids )
...
+    }
+
+    @Override
+    protected Relationship postProcessObject( Relationship relationship )
+    {
+        return (relationship == null || relationship.isDeleted()) ? null : relationship;
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java:324]: I assume this removes deleted rels before they are returned from the store methods?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java [hunk 7]
```diff
@@ -372,9 +373,14 @@ public Enrollment getEnrollment( User user, ProgramInstance programInstance, Tra
         {
             for ( RelationshipItem relationshipItem : programInstance.getRelationshipItems() )
             {
-                enrollment.getRelationships()
-                    .add( relationshipService.getRelationship( relationshipItem.getRelationship(),
-                        RelationshipParams.FALSE, user ) );
+                org.hisp.dhis.relationship.Relationship daoRelationship = relationshipItem.getRelationship();
+                if ( trackerAccessManager.canRead( user, daoRelationship ).isEmpty()
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:377]: what does `canRead().isEmpty()` mean?
should this not just be `canRead()`, so if the user is allowed to read the rel we fetch more details on it

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/persister/RelationshipPersister.java [hunk 8]
```diff
@@ -101,7 +101,7 @@ protected boolean isUpdatable()
     @Override
     protected boolean isNew( TrackerPreheat preheat, Relationship trackerDto )
     {
-        return preheat.getRelationship( trackerDto ) == null;
+        return preheat.getRelationship( trackerDto.getRelationship() ) == null;
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/persister/RelationshipPersister.java:104]: why is that needed? didnt this work before

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 9]
```diff
@@ -631,18 +632,17 @@ public Optional<TrackedEntityComment> getNote( String uid )
         return Optional.ofNullable( notes.get( uid ) );
     }
 
-    public Relationship getRelationship( org.hisp.dhis.tracker.domain.Relationship relationship )
+    public Relationship getRelationship( String relationshipUid )
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:635]: did you have to make this param change?

if so, I would still leave a `public Relationship getRelationship( org.hisp.dhis.tracker.domain.Relationship relationship )` in there that then calls `getRelationship( rel.getRelationship() )`

this would client code nicer to read

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 10]
```diff
@@ -631,18 +632,17 @@ public Optional<TrackedEntityComment> getNote( String uid )
...
     {
-        RelationshipType relationshipType = get( RelationshipType.class, relationship.getRelationshipType() );
+        return relationships.get( relationshipUid );
+    }
 
-        if ( relationship.getUid() != null && relationships.containsKey( relationship.getUid() ) )
-        {
-            return relationships.get( relationship.getUid() );
-        }
+    public Relationship getRelationshipByKey( org.hisp.dhis.tracker.domain.Relationship relationship )
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:640]: could this just be `getRelationshipByKey(Rel...)` ?
What is the meaning of `Key` here and in the other APIs you added?

**Review by teleivo — COMMENTED** [review]
I have added some questions 😄 Things I'd like to understand before ✅

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java (hunk 5 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java:318]: All methods using HibernateGenericStore or HibernateIdentifiableObjectStore methods to retrieve objects from DB

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java (hunk 6 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java:324]: Yes, this is used by get method.
I am really not sure why here is not used the preProcessPredicates as well but this is not part of this PR

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java (hunk 7 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:377]: I agree this api is confusing, apparently canRead returns a list of errors and if there is none then it is ok to read.
I am just using the API here, I am not going down this api, even more as it is part of the old importer

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerErrorCode.java [hunk 11]
```diff
@@ -159,6 +159,8 @@
     E4014( "Relationship Type `{0}` constraint requires a Tracked Entity having type `{1}` but `{2}` was found." ),
     E4015( "Relationship: `{0}`, already exists." ),
     E4016( "Relationship: `{0}`, do not exist." ),
+    E4017( "Relationship: `{0}`, is already deleted and cannot be modified." ),
+    E4018( "Relationship: `{0}`, a relationship linking {1}: `{2}` to {3}: `{4}` already exists." ),
```
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerErrorCode.java:163]: Done

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java (hunk 1 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java:96]: I will actually rename it to duplicatedRelationship

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java (hunk 3 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java:516]: Done

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java (hunk 4 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHookTest.java:514]: Removed

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHook.java [hunk 12]
```diff
@@ -137,14 +138,19 @@ public void validateRelationship( ValidationErrorReporter reporter, Relationship
     {
         TrackerBundle bundle = reporter.getBundle();
         TrackerPreheat preheat = bundle.getPreheat();
-        org.hisp.dhis.relationship.Relationship existingRelationship = preheat.getRelationship(
-            relationship );
+        org.hisp.dhis.relationship.Relationship existingRelationship = preheat
+            .getRelationship( relationship.getUid() );
         TrackerImportStrategy importStrategy = bundle.getStrategy( relationship );
 
+        if ( existingRelationship != null && existingRelationship.isDeleted() )
```
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/PreCheckExistenceValidationHook.java:145]: Done

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java (hunk 9 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:633]: I added the method

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java (hunk 10 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java]: Now it should be clearer.
I separated completely the concept of preheated relationships and duplicated relationships.
The first ones are defined only by the UID, the second ones are defined by the key, that is a combination of the relationship type, from, to and the bidirectional flag

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/TrackerBundle.java [hunk 13]
```diff
@@ -262,4 +262,9 @@ public ProgramStageInstance getProgramStageInstance( String event )
     {
         return getPreheat().getEvent( event );
     }
+
+    public org.hisp.dhis.relationship.Relationship getRelationship( String relationship )
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/TrackerBundle.java:266]: Can you please move this into the preheat. The implementation shows that it actually belongs in there as its just delegating to it. If you have the bundle, you also have the preheat. So this should easy for client code.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 14]
```diff
@@ -267,6 +267,15 @@ public MetadataIdentifier getCategoryOptionComboIdentifier( CategoryCombo catego
     @Getter
     private final Map<String, Relationship> relationships = new HashMap<>();
 
+    /**
+     * Internal map of all relationships in the DB that are duplications of
+     * relationships present in the payload. Two relationship are duplicated if
+     * they have the same relationshipType, and the same relationshipItems, from
+     * and to.
+     */
+    @Getter
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:276]: Do you need the getter? I am asking because I removed getters in the preheat, where I could so we do not expose our mutable collections. I then looked for what client code needed and added these methods instead.

Ideally we would have a way of enforcing that only in the preheat phase can you add things to the preheat (maybe via an interface). All other phases should only get access to immutable data/methods returning immutable data.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 15]
```diff
@@ -267,6 +267,15 @@ public MetadataIdentifier getCategoryOptionComboIdentifier( CategoryCombo catego
...
+    private final Map<String, Relationship> duplicatedRelationships = new HashMap<>();
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:277]: Since English is not my first language take my gut feeling as that 😋 
`duplicateRelationships` feels more correct to me

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 16]
```diff
@@ -267,6 +267,15 @@ public MetadataIdentifier getCategoryOptionComboIdentifier( CategoryCombo catego
     @Getter
     private final Map<String, Relationship> relationships = new HashMap<>();
 
+    /**
+     * Internal map of all relationships in the DB that are duplications of
+     * relationships present in the payload. Two relationship are duplicated if
+     * they have the same relationshipType, and the same relationshipItems, from
+     * and to.
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:274]: can you please add what the key represents, since its a String its always useful, at least to me 😄

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java [hunk 17]
```diff
@@ -93,10 +93,10 @@ public void validateRelationship( ValidationErrorReporter reporter, Relationship
     private void validateDuplication( ValidationErrorReporter reporter, Relationship relationship,
         TrackerBundle bundle )
     {
-        org.hisp.dhis.relationship.Relationship relationshipUsingKey = bundle.getPreheat()
-            .getRelationshipByKey( relationship );
+        org.hisp.dhis.relationship.Relationship duplicateRelationship = bundle.getPreheat()
+            .getDuplicatedRelationship( relationship );
 
-        if ( Objects.nonNull( relationshipUsingKey ) )
+        if ( Objects.nonNull( duplicateRelationship ) )
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java:99]: so my point about not adding a getter would be to add something like `preheat.isDuplicate(Relationship)` instead

**Review by teleivo — APPROVED** [review]
nice 😄 thanks a lot for the changes and make the test strict!
The only thing I would definitely want to change is removing the getter TrackerPreheat.getDuplicateRel... and adding a helper like `isDuplicate(Relationship)` instead.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/TrackerBundle.java (hunk 13 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/TrackerBundle.java:266]: I put it here because it make sense to treat Relationship as any other entity.
We either remove all the others delegation for TEI, Enrollment or Event or we maintain all of them.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java (hunk 14 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java]: Removed

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java (hunk 16 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java]: Done

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java (hunk 17 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java]: Done

**Review by gnespolino — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java (hunk 15 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:278]: Done
